core/Segment.php

```diff
@@ -252,6 +284,51 @@ private function getExpressionsWithUnionsResolved($expressions)
         return $expressionsWithUnions;
     }
 
+    private function isVisitSegment($name)
+    {
+        $availableSegment = $this->getSegmentByName($name);
+
+        if (!empty($availableSegment['unionOfSegments'])) {
+            foreach ($availableSegment['unionOfSegments'] as $segmentNameOfUnion) {
+                $unionSegment = $this->getSegmentByName($segmentNameOfUnion);
+                if (strpos($unionSegment['sqlSegment'], 'log_visit.') === 0) {
+                    return true;
+                }
+            }
+        } else if (strpos($availableSegment['sqlSegment'], 'log_visit.') === 0) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private function doesSegmentNeedSubquery($operator, $segmentName)
+    {
+        $requiresSubQuery = in_array($operator, [
+                SegmentExpression::MATCH_DOES_NOT_CONTAIN,
+                SegmentExpression::MATCH_NOT_EQUAL
+            ]) && !$this->isVisitSegment($segmentName);
+
+        if ($requiresSubQuery && empty($this->startDate)) {
+            $e = new Exception();
+            Log::warning("Avoiding segment subquery due to missing start date. Please ensure a start date is set when initializing a segment if it's used to build a query. Stacktrace:\n" . $e->getTraceAsString());
```

Comment:
I guess you mean when start and end date is empty?

created follow up issue so we don't forget to act on it https://github.com/matomo-org/matomo/issues/16200

Reply:
sorry just realise shouldn't have merged this yet as it needs to also check if `endDate` is empty @sgiehl ?

Reply:
created https://github.com/matomo-org/matomo/pull/16201